### ISSUE TITLE
Error loading kubeconfig || using kind on macOS

### ISSUE THREAD

**Issue by Dortmund-Reus** [krustlet/krustlet#637]
![image](https://user-images.githubusercontent.com/30483659/124068178-21e6db00-da6d-11eb-84a4-5c20032c1c8b.png)

![image](https://user-images.githubusercontent.com/30483659/124066600-8d2fad80-da6b-11eb-9540-bc9a7dafaa56.png)
Why does it say there is no such file?
![image](https://user-images.githubusercontent.com/30483659/124066783-e0a1fb80-da6b-11eb-9b4d-d0af22acf302.png)
here is my bootstrap.conf:
```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUM1ekNDQWMrZ0F3SUJBZ0lCQURBTkJna3Foa2lHOXcwQkFRc0ZBREFWTVJNd0VRWURWUVFERXdwcmRXSmwKY201bGRHVnpNQjRYRFRJeE1EY3dNVEF6TXpnd04xb1hEVE14TURZeU9UQXpNemd3TjFvd0ZURVRNQkVHQTFVRQpBeE1LYTNWaVpYSnVaWFJsY3pDQ0FTSXdEUVlKS29aSWh2Y05BUUVCQlFBRGdnRVBBRENDQVFvQ2dnRUJBTWVnCjdrcU5GV0graTlSTFpEclc3eGhuQ1JaSXBuZXZIWVNMdFNKdUpoTjJsVFRFMFZGUDBMNXhUMVI3bW1uQTJNTHoKS016My8rNVViUWFSUnk2MWdmdzlKczVzaDRuN09xZHh0ZGg4WmlHNkJEaEVmcllJaUdWRDFpNFg3YVBESFo2TgpLS0oyVDdPakoxVzlyQ2d4YUVmQXllVEhsRzBCRkpGbml2OUxOOXFyckxXK3AzYStBNHdzRGVhUHdiSHZUU2V3CjhHV1IzOHkvWG9EZkdreUhWTWFkM1VaSFpMelM1MlBtODZQOEM4ZFcrVkszNlhVNG9YaGlHdW0relVPWGovL20KRUFRZXREV3BlYzMvdEhIbkppQXJOajhLekVhUFVtblRka1YvS1ZZeFlYVVpCR1p3SnhXeWxNc1M0TTQ3RFE1UQpvcXFNMnh2V3c1S21SYkJyTE1VQ0F3RUFBYU5DTUVBd0RnWURWUjBQQVFIL0JBUURBZ0trTUE4R0ExVWRFd0VCCi93UUZNQU1CQWY4d0hRWURWUjBPQkJZRUZOa01QeW5EY1pzSUFnVGRwRTF0VzZ4M2lNcHBNQTBHQ1NxR1NJYjMKRFFFQkN3VUFBNElCQVFBUVNISkRISytITUxNdVI0UWZBSHV1dmlwci9NV1hETTBzeGEySFV0QWlVaHB0ZUYxVwpiRWhuZExDMDFtUzljbUhGYWdrMkhjQ0J4a29JTDZ6NXE4MlNUZWRtTlVmRlhRY2UvRE5sZmQ2dXZWeTVKbU9xClFoRjQ1SXduYmRjdnRwSG9xVlE4NHMvQkVJdkdGcEFoNjZNRllsMGdKR20xNHNwWGlkeGJJOEE3ZHE5VGxzV0IKYUtGVWczOHRBZ2pMYndOOTNvNDhPYTh5WDlGQ2N4QkVJeDdmTUg1ZU12bXp6KzUwTWs0U0szbDNoVm9PNnBiSwpMTzhQQ1hkZG1DejRzSTR5ZzZvY3lsNUpCcUNQRHlDNUZXbGJtemxJeTNNZFBodTNKWHlzL1BZbWFQWlI1YTlKClhPeW15cnpmM0U1Mm1JS3BLZDBDMk8wbjl4Z3FDNFcwMEpBWgotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg==
    server: https://127.0.0.1:63598
  name: kind-kind
contexts:
- context:
    cluster: kind-kind
    namespace: kube-system
    user: tls-bootstrap-token-user
  name: tls-bootstrap-token-user@kubernetes
current-context: tls-bootstrap-token-user@kubernetes
kind: Config
preferences: {}
users:
- name: tls-bootstrap-token-user
  user:
    token: tfbgh6.icn3lrnn565zuq2b
```
And I want to know how can I get the kubeconfig file?
![image](https://user-images.githubusercontent.com/30483659/124067132-6cb42300-da6c-11eb-91e5-bf5b1b2412ab.png)

**Comment by bacongobbler** [comment]
Out of curiosity, can you try using either $HOME or an absolute path (/home/myuser) instead of the tilde (~)?

**Comment by Dortmund-Reus** [comment]
> Out of curiosity, can you try using either $HOME or an absolute path (/home/myuser) instead of the tilde (~)?

![image](https://user-images.githubusercontent.com/30483659/124067192-8f463c00-da6c-11eb-9c93-1b22d8528414.png)
doesn't work for me.

**Comment by bacongobbler** [comment]
Okay, so the first error is no longer present. However the new error would suggest a kubernetes cluster is not listening on port 63598. Can you please check and confirm whether you have the correct kubeconfig file to connect to your cluster, and can you please check if kubernetes is listening on localhost, port 63598?

How did you set up your kubernetes cluster? what steps were you following?

**Comment by Dortmund-Reus** [comment]
I was following this document:
https://github.com/deislabs/krustlet/blob/main/docs/howto/krustlet-on-kind.md

**Comment by Dortmund-Reus** [comment]
> Okay, so the first error is no longer present. However the new error would suggest a kubernetes cluster is not listening on port 63598. Can you please check and confirm whether you have the correct kubeconfig file to connect to your cluster, and can you please check if kubernetes is listening on localhost, port 63598?
> 
> How did you set up your kubernetes cluster? what steps were you following?

I was following this tutorial:
https://github.com/deislabs/krustlet/blob/main/docs/howto/krustlet-on-kind.md

**Comment by Dortmund-Reus** [comment]
> Okay, so the first error is no longer present. However the new error would suggest a kubernetes cluster is not listening on port 63598. Can you please check and confirm whether you have the correct kubeconfig file to connect to your cluster, and can you please check if kubernetes is listening on localhost, port 63598?
> 
> How did you set up your kubernetes cluster? what steps were you following?

I don't know whether I should use the default kubeconfig which exists as ~/.kube/config. Or I should use ~/.krustlet/bootstrap.conf instead?

**Comment by bacongobbler** [comment]
The documentation suggests a third file: ~/.krustlet/config/kubeconfig. I think @thomastaylor312 may know more on the specifics.

https://github.com/deislabs/krustlet/blob/main/docs/howto/bootstrapping.md#generating-the-kubeconfig
https://github.com/deislabs/krustlet/blob/main/docs/howto/bootstrapping.md#running-krustlet

> A couple important notes here. KUBECONFIG should almost always be set, especially in developer/local machine situations. During the bootstrap process, Krustlet will generate a kubeconfig with the credentials it obtains during the bootstrapping process and write it out to the specified location in KUBECONFIG. If a kubeconfig already exists there, it will be loaded and skip the bootstrapping process. A similar process occurs during the bootstrapping of the serving certificates, they will be written out to the paths specified by --cert-file (default $KRUSTLET_DATA_DIR/config/krustlet.crt) and --private-key-file (default $KRUSTLET_DATA_DIR/config/krustlet.key). If they already exist, then they will be loaded and bootstrapping skipped.

For now I'd try using the default kubeconfig file at ~/.kube/config. The documentation suggests bootstrap.conf should be used as input to the `--bootstrap-file` flag.

**Comment by Dortmund-Reus** [comment]
Thank you.Here is what I did:
I made a copy of bootstrap.conf and rename it as kubeconfig. 
Then I run this command:
```
kind create cluster
```
It changes the kubeconfig file and now it looks like this:
```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUM1ekNDQWMrZ0F3SUJBZ0lCQURBTkJna3Foa2lHOXcwQkFRc0ZBREFWTVJNd0VRWURWUVFERXdwcmRXSmwKY201bGRHVnpNQjRYRFRJeE1EY3dNVEEyTVRrek9Gb1hEVE14TURZeU9UQTJNVGt6T0Zvd0ZURVRNQkVHQTFVRQpBeE1LYTNWaVpYSnVaWFJsY3pDQ0FTSXdEUVlKS29aSWh2Y05BUUVCQlFBRGdnRVBBRENDQVFvQ2dnRUJBS240Cm9RWFJ1ZVFhenJnRTFMYVUybVVhTUovUm9zejVyUTFhZ3o1K2xxWkE0VkNKOElOZFBqMENjeGZBQlkwc2xnYkcKamVXNWQxcE5vL3JIWmZnUDVYNjRzZWE2aGtmTzRhV29WYS9FQXZMUW5PQm80NGNTQnJ5b0VqcmNzQU1DNnRoVQpLeG1XcmptNW42ZFM4a3R6MStVSDZyendLMnphUUh4NXI3NW93V09ZOTNyQVNySXJrT0lCNEU1WldnVkFkTTNmClU3UlM4N1dVbHV5M2U3ckdZU1h5WHpvclUycllubENCNTRROGlsa0w0L3BXOThBbDVqL05FbGpWWEl3dVhuRWQKd0J5dkR4bFYyMThLOTEyUjJFY2d1SFNTQXFOVmRmaDJ6V1Z3TVhJTDJ1bm44cDRuZ0ZhNHJFa0Q4OXpsb1llVQpuN01LVVRuS1ZmL096NmdxNHJVQ0F3RUFBYU5DTUVBd0RnWURWUjBQQVFIL0JBUURBZ0trTUE4R0ExVWRFd0VCCi93UUZNQU1CQWY4d0hRWURWUjBPQkJZRUZKYysrcVlZOW9wbHFEeWZHamkrWWZEOGFpNDVNQTBHQ1NxR1NJYjMKRFFFQkN3VUFBNElCQVFCNzJ4VmJsTXl1TU5DbFhFVkF0R0FiZHZtMGhOR0xKdHAwVUhGQkhwVTd6aDk2L3B5dgpyOHFnd0RYbkRKbkdsaHpYcFZYUm5oQU8xNWFLaGwxL0sxSHBBand5cy9iRWRBK1VoK2t6QmxHcWFQMlZhQnlUCm8xbmhZd2gxNktGbnhMdnQzYWxHYytaTXpxeVp0Z2lBOWMxNUJTSS9Ja1k1Y0tzMndOSGoyc3hKT2dBRWNqNGEKSUx5QjFHOGhhbXNmR1VFRWdJNDJDTGFNV3ZuRTNwWXNSMkZ5VWZzdWRFNU5oeWZPY0FZc0kvbUlGaEExYm9EegpOcEN1K0twRnBrWXZMeWU0NWs5eXliVzNjcUh3TWI5bG96MFVZMjVjNHpmNEwrZkdOWmw4V1hmb3dQKzAyL091CjZqU2VZSURxRWdyczlXS1VHUXFVMTNSSWhZS3JJQ2lRbmRHVAotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg==
    server: https://127.0.0.1:61839
  name: kind-kind
contexts:
- context:
    cluster: kind-kind
    namespace: kube-system
    user: tls-bootstrap-token-user
  name: tls-bootstrap-token-user@kubernetes
- context:
    cluster: kind-kind
    user: kind-kind
  name: kind-kind
current-context: kind-kind
kind: Config
preferences: {}
users:
- name: tls-bootstrap-token-user
  user:
    token: ezwg4h.lvljr0o40crmtp1x
- name: kind-kind
  user:
    client-certificate-data: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURJRENDQWdpZ0F3SUJBZ0lITTdmQW5XdWgvREFOQmdrcWhraUc5dzBCQVFzRkFEQVZNUk13RVFZRFZRUUQKRXdwcmRXSmxjbTVsZEdWek1CNFhEVEl4TURjd01UQTJNVGt6T0ZvWERUSXlNRGN3TVRBMk1UazBNVm93TkRFWApNQlVHQTFVRUNoTU9jM2x6ZEdWdE9tMWhjM1JsY25NeEdUQVhCZ05WQkFNVEVHdDFZbVZ5Ym1WMFpYTXRZV1J0CmFXNHdnZ0VpTUEwR0NTcUdTSWIzRFFFQkFRVUFBNElCRHdBd2dnRUtBb0lCQVFDMkozWGgzcVdTSjFuVWZ2UnEKUHdYSURWUUJzMFlrMjBVTnhxaldqUmpVODdvUGVaSDRCVlRxWGZRd1FQaHJJMGgxZDdIZyt2aXlaaEtaY2V5SQpOckVuLzJ0WW9nK1FzcC9ERXpsQVkwbXR4NXZQREtSV0hBZ3J0Z2kxUGhYL3c5ODBoampaTUdEK0VoUlNBbUYxCjlCTk9lYjRiSzVqb2FUS0dCcXdFOG8ySUl2OHdGalpvMVVEK296aThhMU1ZOXRQWG9nQTlwWXZXUGQvcld2cHgKYkZRVW5ZeHVqcm96YkhQQjJaRklUYy9TNjZiYTluTThNWGdDSmUrdmJzKzMxdnpQeE81UEdBRys2SWw2SlBKawptYy9IUlR3RDF3SEh6NmdwNE1qRHdaL05neVdIWFUwL1hxdW4wK3FuUXRSeTVLSXFIQ0gzOGJPMkFlaDRPMm02CmNnQlBBZ01CQUFHalZqQlVNQTRHQTFVZER3RUIvd1FFQXdJRm9EQVRCZ05WSFNVRUREQUtCZ2dyQmdFRkJRY0QKQWpBTUJnTlZIUk1CQWY4RUFqQUFNQjhHQTFVZEl3UVlNQmFBRkpjKytxWVk5b3BscUR5ZkdqaStZZkQ4YWk0NQpNQTBHQ1NxR1NJYjNEUUVCQ3dVQUE0SUJBUUJMZjdHNFpOK0FVOFg1bXdENzFyaHVOUk8zcHhRUTY0WTNDMGtlCjgzUFRPL1dBcXpka2YvY3JTZDhGQ0pvdmo4OTEzS08vR0pmR3BiYmxobFd2a3RibGpaUGhYNjFGaHVlK2VIdFQKU3FnaUY5MzNhTmRnamloS3UzbWRzMHhkSkRsZ2h6Q3RnQmxFTEhyS1NsM09UNjJjN2EraEJZZEZVSDNDaVZOMwo1VjUrakwxVS9NMExOMGxSLzBncVYwREM1Y2U0dW5HMzBoWUQ2bURJbGYxb2dXRGc0Q3cvWmJaTEhZbUMwN2Q2CjVzdXB3Rk5UbSs4clZTZ09pNnJHd09rV0gzK3lzQnUrWFVvTVpSR0hVczVERmxXeWN2Z3FYUjVyQXhKbUFkd3gKRUw0Z0ZjVzJpajdYRzA0dFRkU3pGSk5nYnpQWHRBK0NVL1NKRXhaOHk3OVBlR2kwCi0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K
    client-key-data: LS0tLS1CRUdJTiBSU0EgUFJJVkFURSBLRVktLS0tLQpNSUlFcFFJQkFBS0NBUUVBdGlkMTRkNmxraWRaMUg3MGFqOEZ5QTFVQWJOR0pOdEZEY2FvMW8wWTFQTzZEM21SCitBVlU2bDMwTUVENGF5TklkWGV4NFByNHNtWVNtWEhzaURheEovOXJXS0lQa0xLZnd4TTVRR05KcmNlYnp3eWsKVmh3SUs3WUl0VDRWLzhQZk5JWTQyVEJnL2hJVVVnSmhkZlFUVG5tK0d5dVk2R2t5aGdhc0JQS05pQ0wvTUJZMgphTlZBL3FNNHZHdFRHUGJUMTZJQVBhV0wxajNmNjFyNmNXeFVGSjJNYm82Nk0yeHp3ZG1SU0UzUDB1dW0ydlp6ClBERjRBaVh2cjI3UHQ5Yjh6OFR1VHhnQnZ1aUplaVR5WkpuUHgwVThBOWNCeDgrb0tlREl3OEdmellNbGgxMU4KUDE2cnA5UHFwMExVY3VTaUtod2g5L0d6dGdIb2VEdHB1bklBVHdJREFRQUJBb0lCQURWRjN5MGE3N09KL0xTdwpHSTJHTnd1bUVqNlpEdGV5cFVSOUtGM3ZVQ2JmelRlL1oxaHhQcjk0a1BVWDIwK0NOeXZid05NNnMwZUJvTWp0Cno2NmF5RDhaYld5em9QZkJ0Z1hZZG5NRlZzUVRUOWRETEVEZTIramt5VEV3b2tCMkthNXozdVp1UnIrbnkvZWMKS0tMVnFsdldNU3BER3VwRjhmZk90YnFRc2FIeGsvazNuL2RRQTkvRW5BLzRXWDBnNjBiV2JIREUyZmVaK3hHagpCRjM2TEkycFFFSlJSQkg4V0hYaVRTdXZnSEpUUGpuTHlIQkJyMERSVjRGcTc1T05CQnlRWVFncEJYZlI4YUcwCjZhMFhSZzFYRUdzbDFJU1hmbTJFZ09WcUFNV1hSRmdxZUMxNWF6UnRzengyVnpoOHk1Zm0yejEvNERyVEkwYnMKWktoU2hoRUNnWUVBMC9iZjNlNW1FL2l1bTA3ckpndXdXKzl3ZGQ0ZGt2R2Zla0FncTJFSkpWamlGUTBqV0FlNQp5TlpGdTRtMzFMUHkybllacnBIczR3dGwrM08rMENlSTI1QlkzM2I0Mk5IeWRkYTVTQ0ZMNGhtMmN4MTJoaW9LCk5tUTRkRWpLd3BIM1Y5VFlyaDJRYllYOG5FWHJIb2tSbzlGc1dZZHVhV1c2aUE2UUVldmRDL2NDZ1lFQTIvOG8KYVoxK3FxczAyUDB5RElYZEh6WU5KQk1rZ3ZBR2xLSjZicTJ0WHFHVlVtT3UyckhLSFljSTFkYVFsVTdzMHVYaQpUdFhjZG5wbEV0S0FOYVp4SjVJazk4Vk5UNUZITmN1SkhFL3RRUEk3OXpRTndpd0QvZDIyai9HUnkyOGdxUVBqCkZnZk9hNjNlb1VEOGFVYjc3Zkx1czlwTENZYTIwZlFxUXBNQ3FHa0NnWUVBejdsbTF2WGxMN2xNVklQWXl1WTEKcFVOWDhRc0Y3SlhDZG1vV0doMzBlcGRLcU5yQlFtbi8xaVMzSGZ6dkhSSms4MWM5LytpZVFEdFJiRVdZUUNJZQpheDVoR2s1YnZSQ3pEbXlxRTFnZUo2eVBNdTB5MUhTcW9YZENKQ09hSVdndS8vSmxqdTBHTjI5UmlreXpDSy9ICmk4V1c2MzVmNUFaLzhRcGtna3VseXkwQ2dZRUFzOFdVRG9vbW9leDVxUDl1VUpGckdrQWZDT1dqcG1QaDNML1EKOEhSNEFhWmswUWlvTUFtc3NlaFFyZjFCOU9DajRjcjgzWk1oTHFmaHhEbkRqeFBOejVvN0gwVlpZYlhEelEvZgpvUVVuU1RVcEQxNjRCMzNVWFRVNkkvSXhRMTR5QUNzdS9DZXc1SWtXTGd0ZXQrRExxYXdrcDArd2xiNUIwdHVGClZSV3d6RkVDZ1lFQWxzTkQzNlhBeGV1eXk1NUpIK1RtMkJwSW9HRXdJNWFsZ3BGVFNESUM1TGI3RXlGOVA4R2YKbmJDMERmbTNCbmRyYmtmR0ZSS0V3QVM1dEVCTW9JeEJkMEZ5WkRJMFBKanRJNGRLWDdteHhmb1gzZktEWGF5UQpkdkhnWFg4TTRGTTZWVGFLdC9CWmJjR0Z4VzZ2a3FBN1hUSFNVTUlram4yeFZ1VlRrM2VZZ1hZPQotLS0tLUVORCBSU0EgUFJJVkFURSBLRVktLS0tLQo=

```

**Comment by thomastaylor312** [comment]
So here is how the logic works:

1. On startup, Krustlet looks for the file specified by `KUBECONFIG` 
2. If `KUBECONFIG` exists, it assumes that bootstrapping is complete and uses that kubeconfig file to connect
3. If `KUBECONFIG` does not exist, it will load the bootstrap file specified by `--bootstrap-file` (or the default)
4. Using the bootstrap file, it will generate the kubeconfig using the certificates provided by the api and write the file to `KUBECONFIG`

So basically, your `KUBECONFIG` variable on first startup is telling krustlet where you want to write out the generated kubeconfig file.

The last error you attached a screenshot of with the decoding error is something I haven't seen before. However, if you set `KUBECONFIG` properly along with a valid `--bootstrap-file`, does the error still happen?

**Comment by Dortmund-Reus** [comment]
> So here is how the logic works:
> 
> 1. On startup, Krustlet looks for the file specified by `KUBECONFIG`
> 2. If `KUBECONFIG` exists, it assumes that bootstrapping is complete and uses that kubeconfig file to connect
> 3. If `KUBECONFIG` does not exist, it will load the bootstrap file specified by `--bootstrap-file` (or the default)
> 4. Using the bootstrap file, it will generate the kubeconfig using the certificates provided by the api and write the file to `KUBECONFIG`
> 
> So basically, your `KUBECONFIG` variable on first startup is telling krustlet where you want to write out the generated kubeconfig file.
> 
> The last error you attached a screenshot of with the decoding error is something I haven't seen before. However, if you set `KUBECONFIG` properly along with a valid `--bootstrap-file`, does the error still happen?
I don't know if I understand it correctly. So if the file specified by KUBECONFIG(~/.krustlet/config/kubeconfig) does not exist, the krustlet-wasi command will load the bootstrap file specified by --bootstrap-file, and then generate a new file kubeconfig at ~/.krustlet/config?
```
export KUBECONFIG=~/.krustlet/config/kubeconfig
```

**Comment by thomastaylor312** [comment]
With `export KUBECONFIG=~/.krustlet/config/kubeconfig`, if it does not exist, `krustlet-wasi` will load the `--bootstrap-file` to do the bootstrapping process and generate a valid "real" kubeconfig at `~/.krustlet/config/kubeconfig`. Then it will use that new kubeconfig file to authenticate to the cluster

**Comment by Dortmund-Reus** [comment]
> With `export KUBECONFIG=~/.krustlet/config/kubeconfig`, if it does not exist, `krustlet-wasi` will load the `--bootstrap-file` to do the bootstrapping process and generate a valid "real" kubeconfig at `~/.krustlet/config/kubeconfig`. Then it will use that new kubeconfig file to authenticate to the cluster

OK.I delete the KUBECONFIG file,and run these commands:
```
kind create cluster
./docs/howto/assets/bootstrap.sh
export KUBECONFIG=~/.krustlet/config/kubeconfig
krustlet-wasi --node-ip 192.168.50.112 --bootstrap-file=$HOME/.krustlet/config/bootstrap.conf
kubectl certificate approve MBP-de-Elio-tls
```

![image](https://user-images.githubusercontent.com/30483659/124209804-cd029d80-db1c-11eb-97cb-0b09c52595f7.png)

![image](https://user-images.githubusercontent.com/30483659/124209820-d429ab80-db1c-11eb-92a6-5ac9a33a8176.png)

**Comment by thomastaylor312** [comment]
Does the node crash after those errors? It looks like you have a pod that is scheduling to the node that has an image it doesn't know how to understand (happening in this logic [here](https://github.com/deislabs/krustlet/blob/main/crates/kubelet/src/state/common/image_pull.rs)). 

2 questions here:
1. What is your k8s version that kind is running?
2. What is the output of `kubectl get pods --all-namespaces -o wide` once that error occurs

Basically it looks like you are bootstrapping just fine, but an error is occurring once pods get scheduled to the node

**Comment by Dortmund-Reus** [comment]
> Does the node crash after those errors? It looks like you have a pod that is scheduling to the node that has an image it doesn't know how to understand (happening in this logic [here](https://github.com/deislabs/krustlet/blob/main/crates/kubelet/src/state/common/image_pull.rs)).
> 
> 2 questions here:
> 
> 1. What is your k8s version that kind is running?
> 2. What is the output of `kubectl get pods --all-namespaces -o wide` once that error occurs
> 
> Basically it looks like you are bootstrapping just fine, but an error is occurring once pods get scheduled to the node

![image](https://user-images.githubusercontent.com/30483659/124371544-5351e780-dcb5-11eb-8605-636f6f72b599.png)

![image](https://user-images.githubusercontent.com/30483659/124371548-68c71180-dcb5-11eb-860f-6a0d6d7ff29c.png)

**Comment by thomastaylor312** [comment]
Ok, so I know there are definitely problems with 1.21 right now (at least one we have an issue for). I am going to work on that issue this week, so that _may_ solve the problem.

As for the IP address part, I'll have to check what happens on my mac

**Comment by Dortmund-Reus** [comment]
> Ok, so I know there are definitely problems with 1.21 right now (at least one we have an issue for). I am going to work on that issue this week, so that _may_ solve the problem.
> 
> As for the IP address part, I'll have to check what happens on my mac

Thank You!

**Comment by bacongobbler** [comment]
Gonna assign this to @thomastaylor312 as I know he's working on some of the issues with 1.21.

Have you re-tried the same steps with Kubernetes 1.20? The AKS deploy script uses 1.20.7 and I tested that last week, so you should have better luck with a slightly older version.

**Comment by bacongobbler** [comment]
https://github.com/deislabs/krustlet/pull/646

**Comment by VishnuJin** [comment]
hey guys, I built the latest tag from source on mac and it works fine for me in KinD, I'm wondering why the IP for the pod is shown as <none> ☹️ !!
<img width="1280" alt="Screenshot 2021-08-05 at 9 37 47 PM" src="https://user-images.githubusercontent.com/45007338/128384452-c5c85a57-c518-43a4-9ab2-e9c9c91103bb.png">
<img width="1280" alt="Screenshot 2021-08-05 at 9 41 40 PM" src="https://user-images.githubusercontent.com/45007338/128384481-f86d8a68-d560-4e66-aa2f-3cd5c039f6c8.png">

Though it works these are the error logs in Krustlet
<img width="1280" alt="Screenshot 2021-08-05 at 9 46 34 PM" src="https://user-images.githubusercontent.com/45007338/128384860-1069d07b-d546-47de-8367-587096b0e401.png">

**Comment by thomastaylor312** [comment]
So the pod doesn't have an IP address because the WASI provider doesn't support inbound networking yet (that functionality is still being worked on in the WASI spec). Currently, wasm modules running in the WASI provider can make outbound http calls.

Those error messages are a known issue as well and they are just noise (particularly on KinD). For some reason the node taints don't repel the kindnet pods and when the wasi-provider tries to pull them, it recognizes that it isn't a wasm module

**Comment by VishnuJin** [comment]
@thomastaylor312 thanks for the clarification 👍

**Comment by thomastaylor312** [comment]
So were there any other problems related to this issue @VishnuJin or can I go ahead and close it?

**Comment by VishnuJin** [comment]
@thomastaylor312 it works fine for me !! I Believe it should fine for @Dortmund-Reus as well ??

**Comment by bacongobbler** [comment]
I'm going to assume this is safe to close at this point. Please let us know if you'd like us to re-open this and re-investigate the issue. Thanks.
